[DO NOT MERGE] DOCS-12770 lifecycle states (#166)
mulesoft/docs-exchange: modules/ROOT/pages/lifecycle.adoc

= Manage Asset Lifecycle States
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images/

In the software development lifecycle, assets are developed, released, and eventually deprecated, and each Exchange asset version moves through the lifecycle states of `development`, `stable`, and `deprecated`.

== Development State

An asset version in the `development` state is in an iterative process of design and development.

If you re-publish an asset version in the `development` state with the same groupId, assetId, and version (GAV) and new metadata, the asset version is overwritten, including this asset version metadata:

* files
* dependencies
* keywords

Changing this information for an asset version in any other state requires creating a new version, which prevents re-publishing or overwriting the asset version.

Overwriting an asset version also updates the search index.

The Exchange user interface and APIs change some asset version information for asset versions in any state. This information is not changed by overwriting an asset version in the `development` state. This information includes:

> * tags

Reviewer: ```suggestion
* asset type
* tags
```

Author: Thank you very much! Added "asset type" to the list.

= Manage Asset Lifecycle States
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images/

In the software development lifecycle, assets are developed, released, and eventually deprecated, and each Exchange asset version moves through the lifecycle states of `development`, `stable`, and `deprecated`.

== Development State

An asset version in the `development` state is in an iterative process of design and development.

If you republish an asset version in the `development` state with the same groupId, assetId, and version (GAV) and new metadata, the asset version is overwritten, including the following version metadata:

* files
* dependencies
* keywords

Changing this information for an asset version in any other state requires creating a new version, which prevents republishing or overwriting the asset version.

Overwriting an asset version also updates the search index.

> The Exchange user interface and APIs change some asset version information for asset versions in any state. This information is not changed by overwriting an asset version in the `development` state. This information includes:

Reviewer 2: This paragraph doesn't make sense to me, I'd just add that there is information that is not overwritten by republishing an asset in development state, and list it

* asset type
* tags
* categories
* custom fields
* documentation pages

Deleting an asset version in the `development` state is always a hard delete, enabling you to reuse the asset's name, groupId, assetId, and version.

The public portal does not show asset versions in the `development` state.

> Anypoint Platform products other than Exchange, such as Anypoint API Designer, Anypoint Flow Designer, or Anypoint Studio, do not consume asset versions in the `development` state.

Reviewer: ```suggestion
Anypoint Platform products other than Exchange, such as Anypoint Flow Designer, or Anypoint Studio, do not consume asset versions in the `development` state.
```

Deleting an asset version in the `stable` state in the first seven days is a hard delete, enabling you to reuse the asset's name, groupId, assetId, and version. Deleting an asset version in the `stable` state after the first seven days is a soft delete, preventing you from reusing the asset's name, groupId, assetId, and version. After a hard delete or a soft delete, an asset cannot be recovered or undeleted.

The Exchange backend and some API responses use the term `published` to mean `stable`.

== Deprecated State

An asset version in the `deprecated` state is similar to an asset version in the `stable` state. The only difference is that the `status` property shows the asset version is deprecated.

== Create an Asset in the Exchange User Interface

. Click *Publish new asset*.
. Set *Lifecycle state* to the *Development* state or the default *Stable* state.
. Set other asset information.
. Click *Publish*.

== State Promotion

These are the three possible state promotions:

. `development` to `stable`
. `stable` to `deprecated`
. `deprecated` to `stable`

Promote an asset version's state either in the asset's Exchange portal or by using the Exchange API.

> In the asset's portal, click the lifecycle state button that displays the current state, and select the new state of *Development*, *Stable*, or *Deprecated*.

Reviewer: ```suggestion
In the asset's documentation page, click the lifecycle state button that displays the current state, and select the new state of *Development*, *Stable*, or *Deprecated*.
```

To use the Exchange API, follow the xref:exchange-api.adoc#promote-lifecycle-state[Promote Lifecycle State] example.

> == Dependencies

Reviewer 3: Can this title be made more robust? What, specifically, about dependencies, or what type of dependencies?
I'd at least add "State" to the title.

Reviewer 3: ...and do we need some text explaining what dependencies are and why your audience should care?

An asset in the `stable` state cannot add an asset in the `development` state as a dependency.

> If an asset in the `development` state depends on one or more other assets in the `development` state, Exchange does not prevent promoting that asset to the `stable` state. However, if a dependency in the `development` state is overwritten, the packages of the asset that was promoted to the `stable` state are not regenerated.

Reviewer 2: This text is not valid.

If an asset in the `development` state depends on one or more other assets in the `development` state, Exchange does not *allow* promoting that asset to the `stable` state, all its dependencies should first be promoted to `stable` first.

Apart from that, if you have an asset in `development` state with dependencies in `development` state and you modify a dependency, the dependent asset will also not be regenerated, it's the user's responsibility to republish the dependent asset once it's dependencies were updated

> An asset in the `stable` state cannot add an asset in the `development` state as a dependency.
> 
> If an asset in the `development` state depends on one or more other assets in the `development` state, Exchange does not allow promoting that asset to the `stable` state. All dependencies of the asset should be promoted to stable first. If an asset in the `development` state has dependencies in `development` state and the dependent is modified, the dependent asset is also not regenerated. It is the asset owner's responsibility to republish the dependent asset once its dependencies are updated.

Reviewer 3: ```suggestion
All dependencies of an asset must be promoted before that asset can be promoted. For example, an asset in the `development` state cannot be promoted to the `stable` state until its dependencies are stable.
+
```
The second half of this is confusing: what are "regeneration" and "republishing" and how do they relate to asset state promotion? What do we mean by "...and the dependent is modified"? Does that mean promoted? Are we saying that the promotion of a dependency requires the asset owner to regenerate/republish the dependency for the promotion to be committed/official?

Reviewer 2: What must be clear is that you might have an asset that has embedded `development` dependencies in its files. If one of those dependencies is modified, the dependent asset will still have a previous version of the dependencies and needs to be republished in order for those files to be updated.

An example would be RAML APIs using RAML fragments published through API Designer

== Create an Asset Using the Exchange API

To create an asset using the Exchange API, follow the xref:exchange-api.adoc#set-lifecycle-state[Set Lifecycle State] example.

== Create an Asset Using Maven

To create an asset using the Exchange Maven Facade API, follow the xref:to-publish-assets-maven.adoc#asset-lifecycle-state[Asset Lifecycle State] example.

> == Search Using the Graph API by Lifecycle State

Reviewer 3: The narrative mentions nothing about a graph; reword for clarity, or add text about a graph named "API by Lifecycle"?
Do we mean that in addition to finding an asset by search/selecting its name, we can also find it by its state? If so, I think that his text might belong in the topic about searching for assets.

To consume an asset and its lifecycle state, or to search for an asset by lifecycle state, follow the xref:to-search-with-graph-api.adoc#asset-lifecycle-state[Asset Lifecycle State] examples.

== See Also

* xref:to-delete-asset.adoc[Delete an Asset]
* xref:exchange-api.adoc[Publish Assets Using the Exchange API]
* xref:to-publish-assets-maven.adoc[Publish Assets Using Maven]
* xref:to-search-with-graph-api.adoc[Search Using the Graph API]

> * https://anypoint.mulesoft.com/graph/api/v2/graphql[Graph API Web Interface]

Reviewer 3: **BREAK to commit**

Reviewer 3: Retry commit; concurrent reviews causing connection issues.
